### PR TITLE
Torch.compile() testing PR

### PR DESCRIPTION
**Pull request by AlphaAtlas**
I have been testing Pytorch's new `torch.compile()` feature in VoltaML. In a nutshell, it sorta works with some quirks and gives a modest performance boost on my laptop RTX 2060, and I am leaving the testing code in this PR along with some notes for anyone else interested in trying... it is not meant to be merged. 

- I am testing on Pytorch 2.1 CUDA 12.1 nightly, with xformers, on Linux.
- Txt2img, img2img and controlnet all work with the default torch.compile config and most options (including LORA) but changing the resolution or batch size forces a lengthy recompile and hits the cache limit after a few changes.
- Setting "dynamic=True" works with almost everything _without recompilation._ Self attention forces recompilation, but works. However, it gives a bizarre syntax error when changing resolution which I am going to try and replicate in `diffusers` and report on the PyTorch repo: [Ne_Error.txt](https://github.com/VoltaML/voltaML-fast-stable-diffusion/files/11447581/Ne_Error.txt)
- The `max-autotune`preset creates a large vram spike, but `max-autotune-no-cudagraphs` does not.
- TomeSD "works," but seems to confuse torch.compile and forces excessive recompilation until inductor hits the cache limit. I intend to test this in `diffusers` and report it upstream.
- The [Hidet backend](https://pytorch.org/blog/introducing-hidet/) is promising, but errors out during compilation. I intend to test this in `diffusers` and report it upstream. 
- I left a default config dict in compile.py. There are more potentially performance enhancing options that are disabled by default, and a disk caching option that needs testing.

### CODE REVIEW

**Comment by AlphaAtlas** [comment]
I just saw that y'all are testing IREE backends... not sure if that will make torch.compile obsolete. 

Also, AITemplate is much faster than torch.compile on my machine, but it requires a batch size of 1 and seems to reduce the quality of SD output.

**Comment by gabe56f** [comment]
IREE is still just an idea that's being discussed, because as I see it right now, it would require drastic changes to diffusers code, which is sadly not in the scope of this project. I'll try to force it to comply, but something that works (as in torch.compile) is far better than something that doesn't.

Weird thing is, I've tried torch.compile myself before tracing was implemented, but I only saw performance regressions, could possibly be something new they did in between 2.0 and 2.1

some context: I use a 3080

**Comment by AlphaAtlas** [comment]
I think the compile needs to called each time an image is generated, especially with dynamic=False. But it doesn't take any time. This could definitely be cleaner though, but lets see if compilation works first...

A Hidet issue mentioned that compilation errors are not necessarily fatal, which I did not know: https://github.com/hidet-org/hidet/issues/202#issuecomment-1528803472

**Comment by AlphaAtlas** [comment]
torch.compile works pretty well for a vanilla diffusers pipeline (until the resolution changes), like this: https://github.com/AlphaAtlas/Diffusion-Compilaton-Testing/blob/main/inductor_test.py

It compiles quickly (less than a minute) with no spam in the log.

But for some reason, the implementation in VoltaML takes forever to compile, fills up the log, and errors out if tomesd is enabled: [tomesd_inductor_voltaml_log.txt](https://github.com/VoltaML/voltaML-fast-stable-diffusion/files/11457151/tomesd_inductor_voltaml_log.txt)

Here is the log of a simple diffusers test: [tomesd_inductor_diffusers_log.txt](https://github.com/VoltaML/voltaML-fast-stable-diffusion/files/11457162/tomesd_inductor_diffusers_log.txt)

The simple test code: https://github.com/AlphaAtlas/Diffusion-Compilaton-Testing/blob/main/tome_inductor_test.py

**Comment by AlphaAtlas** [comment]
Also, torch.compile works in voltaml with tomesd disabled, but still takes forever and fills up the compilation cache.

File: core/inference/compile.py
[the remark this answers is not in the document]
Comment (Stax124) [reply core/inference/compile.py:1]: I think that it would be best we leave it in another file and just import it as a function into core/optimizations

**Comment by AlphaAtlas** [comment]
Going through the non default options one by one, the optimal config for torch.compile seems to be 
```
odict = {
    'dce': True, #dead code elimination, tiny gain
    'permute_fusion': True, #Fx-based linear/matmul/bmm + permute/transpose vertical fusion, small gain
}
```

You can play with the max-autotune thing with some combination of these environment variables:  
```TORCHINDUCTOR_MAX_AUTOTUNE=1 TORCHINDUCTOR_MAX_AUTOTUNE_POINTWISE=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 TORCHINDUCTOR_SEARCH_AUTOTUNE_CACHE=1 TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_AUTOTUNE_IN_SUBPROC=1```

But (on my RTX 2060 machine) no combination seems to be worth it.

**Pull request closed without merging**